Tasks/PackerBuild/src/azureSpnTemplateVariablesProvider.ts

```diff
@@ -23,18 +27,47 @@ export default class AzureSpnTemplateVariablesProvider implements definitions.IT
         if(taskParameters.templateType === constants.TemplateTypeCustom) {
             this._spnVariables = new Map<string, string>();
             return this._spnVariables;
-        } 
+        }
 
         this._spnVariables = new Map<string, string>();
         var connectedService = taskParameters.serviceEndpoint;
-        this._spnVariables.set(constants.TemplateVariableSubscriptionIdName, tl.getEndpointDataParameter(connectedService, "SubscriptionId", true));
+        var subscriptionId: string = tl.getEndpointDataParameter(connectedService, "SubscriptionId", true)
+        this._spnVariables.set(constants.TemplateVariableSubscriptionIdName, subscriptionId);
         this._spnVariables.set(constants.TemplateVariableClientIdName, tl.getEndpointAuthorizationParameter(connectedService, 'serviceprincipalid', false));
         this._spnVariables.set(constants.TemplateVariableClientSecretName, tl.getEndpointAuthorizationParameter(connectedService, 'serviceprincipalkey', false));
         this._spnVariables.set(constants.TemplateVariableTenantIdName, tl.getEndpointAuthorizationParameter(connectedService, 'tenantid', false));
-        this._spnVariables.set(constants.TemplateVariableObjectIdName, tl.getEndpointDataParameter(connectedService, "spnObjectId", true));        
-        
+
+
+        var spnObjectId = tl.getEndpointDataParameter(connectedService, "spnObjectId", true);
+        // if we are creating windows VM and SPN object-id is not available in service endpoint, fetch it from Graph endpoint
+        // NOP for nix
+        if(!spnObjectId && taskParameters.osType.toLowerCase().match(/^win/)) {
+            spnObjectId = await this.getServicePrincipalObjectId(taskParameters.graphCredentials, subscriptionId);
+        }
+
+        this._spnVariables.set(constants.TemplateVariableObjectIdName, spnObjectId);
+
         return this._spnVariables;
     }
 
+    private async getServicePrincipalObjectId(graphCredentials: msRestAzure.ApplicationTokenCredentials, subscriptionId: string): Promise<string> {
+        console.log(tl.loc("FetchingSPNDetailsRemotely", graphCredentials.getClientId()));
+        var client = new azureGraph.GraphManagementClient(graphCredentials, subscriptionId, graphCredentials.activeDirectoryResourceId);
```

Comment:
Did so because base AzureServiceClient requires a non-null subscription value. Will change it to take null value for graph client.